File: sdks/typescript/src/budget.ts
import type { BudgetConfig } from "./types.js";

export class BudgetExceeded extends Error {
  constructor(
    public spent: number,
    public limit: number,
    public scope: string = "run",
  ) {
    super(
      `Budget exceeded (${scope}): spent $${spent.toFixed(4)} of $${limit.toFixed(4)} limit`,
    );
    this.name = "BudgetExceeded";
  }
}

export class Budget {
  private runSpent = 0;
  private daySpent = 0;
  private monthSpent = 0;
  private currentDay: number;
  private currentMonth: number;
  private readonly config: Required<BudgetConfig>;

  constructor(config: BudgetConfig = {}) {
    this.config = {
      maxPerRun: config.maxPerRun ?? Infinity,
      maxPerDay: config.maxPerDay ?? Infinity,
      maxPerMonth: config.maxPerMonth ?? Infinity,
      warnAt: config.warnAt ?? 0.8,
    };
    const now = new Date();
    this.currentDay = now.getUTCDate();
    this.currentMonth = now.getUTCMonth();
  }

  check(proposedCost: number): void {
    this.resetIfNewPeriod();

    if (this.runSpent + proposedCost > this.config.maxPerRun) {
      throw new BudgetExceeded(this.runSpent, this.config.maxPerRun, "run");
    }
    if (this.daySpent + proposedCost > this.config.maxPerDay) {
      throw new BudgetExceeded(this.daySpent, this.config.maxPerDay, "day");
    }
    if (this.monthSpent + proposedCost > this.config.maxPerMonth) {
      throw new BudgetExceeded(
        this.monthSpent,
        this.config.maxPerMonth,
        "month",
      );
    }
  }

  record(cost: number): void {
    this.resetIfNewPeriod();
    this.runSpent += cost;
    this.daySpent += cost;
    this.monthSpent += cost;
  }

  get spent(): number {
    return this.runSpent;
  }

  get remaining(): number | undefined {
    if (this.config.maxPerRun === Infinity) return undefined;
    return Math.max(0, this.config.maxPerRun - this.runSpent);
  }

  private resetIfNewPeriod(): void {
    const now = new Date();
    const day = now.getUTCDate();
    const month = now.getUTCMonth();

    if (day !== this.currentDay) {
      this.daySpent = 0;
      this.currentDay = day;
    }
    if (month !== this.currentMonth) {
      this.monthSpent = 0;
      this.currentMonth = month;
    }
  }
}

File: crates/archon-core/src/budget.rs
use crate::types::BudgetConfig;
use thiserror::Error;

#[derive(Debug, Error)]
pub enum BudgetError {
    #[error("budget exceeded: spent ${spent:.4} of ${limit:.4} limit")]
    Exceeded { spent: f64, limit: f64 },
    #[error("budget warning: spent ${spent:.4} ({pct:.0}% of ${limit:.4})")]
    Warning { spent: f64, limit: f64, pct: f64 },
}

/// Tracks spending and enforces budget limits.
#[derive(Debug, Clone)]
pub struct BudgetTracker {
    config: BudgetConfig,
    run_spent: f64,
    day_spent: f64,
    month_spent: f64,
}

impl BudgetTracker {
    pub fn new(config: BudgetConfig) -> Self {
        Self {
            config,
            run_spent: 0.0,
            day_spent: 0.0,
            month_spent: 0.0,
        }
    }

    /// Check if a proposed cost would exceed any budget limit.
    pub fn check(&self, proposed_cost: f64) -> Result<(), BudgetError> {
        if let Some(limit) = self.config.max_per_run {
            let after = self.run_spent + proposed_cost;
            if after > limit {
                return Err(BudgetError::Exceeded { spent: self.run_spent, limit });
            }
            let pct = after / limit;
            if pct >= self.config.warn_at_fraction {
                tracing::warn!(spent = self.run_spent, limit, pct, "budget warning");
            }
        }
        if let Some(limit) = self.config.max_per_day {
            if self.day_spent + proposed_cost > limit {
                return Err(BudgetError::Exceeded { spent: self.day_spent, limit });
            }
        }
        if let Some(limit) = self.config.max_per_month {
            if self.month_spent + proposed_cost > limit {
                return Err(BudgetError::Exceeded { spent: self.month_spent, limit });
            }
        }
        Ok(())
    }

    /// Record actual spending after a successful LLM call.
    pub fn record(&mut self, cost: f64) {
        self.run_spent += cost;
        self.day_spent += cost;
        self.month_spent += cost;
    }

    pub fn run_spent(&self) -> f64 {
        self.run_spent
    }

    pub fn remaining_run_budget(&self) -> Option<f64> {
        self.config.max_per_run.map(|limit| (limit - self.run_spent).max(0.0))
    }
}
